### PR TITLE
Propose added integration tests

### PR DESCRIPTION
**Pull request by jonah-kohn**
These tests are subject to lots of change via suggestions :)

### CODE REVIEW

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 1]
```diff
@@ -34,6 +34,7 @@
 # Unique ID for this build, can be used as a label for an AI Platform training job.
 # BUILD_ID = os.environ['BUILD_ID']
 
+MACHINES = ["K80", "P100", "V100", "P4", "T4", "TPU_V2", "TPU_V3"]
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: This line has been removed in the most recent commit

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 2]
```diff
@@ -55,8 +56,9 @@ def delete_dir(self, path: Text) -> None:
         if tf.io.gfile.isdir(path):
             tf.io.gfile.rmtree(path)
 
+    #Custom dist strat with 2 workers and reqs
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: All custom dist_strat mwms tests will use 2 workers to provide extra coverage, with 1 worker they become redundant with auto_dist_strat mwms tests.

[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: All comments have been removed.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 3]
```diff
@@ -65,6 +67,273 @@ def test_none_dist_strat_mwms_on_script(self, mock_exit):
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_ctl.py"),
+            distribution_strategy=None,
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    #Custom dist strat with 0 workers and reqs
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_none_dist_strat_mwms_0w(self, mock_exit):
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: Tests have been updated to only use multi-worker strategies with >0 workers.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 4]
```diff
@@ -65,6 +67,273 @@ def test_none_dist_strat_mwms_on_script(self, mock_exit):
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_ctl.py"),
+            distribution_strategy=None,
+            worker_count=0,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    #Auto dist strat with 2 workers and reqs
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_2w(self, mock_exit):
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: All names have been updated to accurately reflect test content.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 5]
```diff
@@ -65,6 +67,273 @@ def test_none_dist_strat_mwms_on_script(self, mock_exit):
...
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    #Auto dist strat with 0 workers and reqs
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_0w(self, mock_exit):
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: All names have been updated to accurately reflect test content.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 6]
```diff
@@ -65,6 +67,273 @@ def test_none_dist_strat_mwms_on_script(self, mock_exit):
...
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=0,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_1x_K80(self, mock_exit):
```
[the remark this answers is not in the document]
Comment (jonah-kohn) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: All tests now use auto chief and worker configs, with the exception of TPU tests.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 7]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+            docker_base_image="tensorflow/tensorflow:latest-gpu",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_none_dist_strat_mwms(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:139]: Can we change this to test_none_dist_strat_multi_worker

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 8]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_ctl.py"),
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:141]: Please pass None to distribution_strategy input here.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 9]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_ows_custom_img(self, mock_exit):
```
[the remark this answers is not in the document]
Comment (pavithrasv) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:97]: Can remove all the other custom image tests.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py (hunk 8 above)
Comment (pavithrasv) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:141]: No need for bucket build, custom img tests here with none dist strat.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 10]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
-            distribution_strategy=None,
             worker_count=2,
             requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+            docker_base_image="tensorflow/tensorflow:latest-gpu",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_tpu_dist_strat_mwms(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:179]: TPU is a distribution strategy just like multi worker strategy. Please rename to test_auto_tpu

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 11]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            chief_config=tfc.COMMON_MACHINE_CONFIGS["CPU"],
+            worker_count=1,
+            worker_config=tfc.COMMON_MACHINE_CONFIGS["TPU"],
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_tpu_dist_strat_mwms_bucket_build(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:190]: remove

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 12]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            chief_config=tfc.COMMON_MACHINE_CONFIGS["CPU"],
+            worker_count=1,
+            worker_config=tfc.COMMON_MACHINE_CONFIGS["TPU"],
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_tpu_dist_strat_mwms_custom_img(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:202]: remove

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 13]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            worker_config=tfc.COMMON_MACHINE_CONFIGS["TPU"],
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_base_image="tensorflow/tensorflow:latest-gpu",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_tpu_dist_strat_mwms_bucket_build_custom_img(self, mock_exit):
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:216]: remove.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 14]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            worker_count=1,
+            worker_config=tfc.COMMON_MACHINE_CONFIGS["TPU"],
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+            docker_base_image="tensorflow/tensorflow:latest-gpu",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_ows_stream_logs(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:227]: test_auto_one_device_stream_logs

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 15]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=0,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            stream_logs=True,
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_ows_job_labels(self, mock_exit):
+        tfc.run(
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:238]: test_auto_one_device_job_labels

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 16]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
             tf.io.gfile.rmtree(path)
 
     @mock.patch.object(sys, "exit", autospec=True)
-    def test_none_dist_strat_mwms_on_script(self, mock_exit):
+    def test_auto_dist_strat_ows(self, mock_exit):
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=0,
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:62]: Please remove worker_count from the tests where worker_count is 0.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 17]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            job_labels={"job": "on_script_tests", "team": "keras"},
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_dist_strat_ows_job_labels_stream_logs(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:247]: remove this and the two tests below.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 18]
```diff
@@ -56,12 +56,228 @@ def delete_dir(self, path: Text) -> None:
...
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_base_image="tensorflow/tensorflow:latest-gpu",
+            job_labels={"job": "on_script_tests", "team": "keras"},
+            stream_logs=True,
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:269]: Please add a test for mirrored_strategy `test_auto_mirrored_strategy`

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavithrasv) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py]: The configuration will be the same in both the use cases.

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 19]
```diff
@@ -56,7 +56,52 @@ def delete_dir(self, path: Text) -> None:
...
+    def test_auto_one_device_strategy_bucket_build(self, mock_exit):
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+            docker_image_bucket_name="TEST_BUCKET",
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_multi_worker(self, mock_exit):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:85]: nit: add _strategy in the end to be consistent with the other tests.

Comment (pavithrasv) [reply src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:85]: for all multi_worker tests

File: src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py [hunk 20]
```diff
@@ -56,7 +56,52 @@ def delete_dir(self, path: Text) -> None:
...
+        tfc.run(
+            entry_point=os.path.join(self.test_data_path, "mnist_example_using_fit.py"),
+            worker_count=1,
+            requirements_txt=os.path.join(self.test_data_path, "requirements.txt"),
+        )
+        mock_exit.assert_called_once_with(0)
+
+    @mock.patch.object(sys, "exit", autospec=True)
+    def test_auto_multi_worker_bucket_build(self, mock_exit):
+        tfc.run(
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/tests/integration/run_on_script_test.py:95]: this test can be removed as well.
